crates/apollo-compiler/src/validation/subscription.rs

```diff
@@ -0,0 +1,57 @@
+use crate::{
+    diagnostics::IntrospectionField, hir::Selection, ApolloDiagnostic, FileId, ValidationDatabase,
+};
+
+pub fn check(db: &dyn ValidationDatabase, file_id: FileId) -> Vec<ApolloDiagnostic> {
+    let mut diagnostics = Vec::new();
+    let subscription_operations = db.upcast().subscription_operations(file_id);
+
+    // Subscription fields must not have an introspection field in the selection set.
+    //
+    // Return an Introspection Field error in case of an introspection field existing as the root field in the set.
+    for op in subscription_operations.iter() {
+        for selection in op.selection_set().selection() {
+            if let Selection::Field(field) = selection {
+                if field.name().starts_with("__") {
+                    diagnostics.push(ApolloDiagnostic::IntrospectionField(IntrospectionField {
+                        field: field.name().into(),
+                        src: db.source_code(op.loc().file_id()),
+                        definition: (field.loc.offset(), field.loc.node_len()).into(),
+                        help: Some(format!(
+                            "`{}` is an introspection field. This is not allowed as the root field for Subscription operations.", field.name()
+                        )),
+                    }));
+                }
+            }
+        }
+    }
+
+    diagnostics
+}
+
+#[cfg(test)]
+
```

Comment:
@lrlna I kept this module-level test in the module, but should it be removed for the sake of convention and consistency?

Reply:
yes, let's keep it these sort of tests in `test_data` directory.